enhancements/insights/conditional-data-gathering.md

```diff
@@ -0,0 +1,231 @@
+to make collection of new conditional data faster. It's NOT about running brand new code,
+but just calling the existing gatherers with different validated parameters, so the operator can't
+collect anything we don't expect.
+
+\* note that rule here and later has nothing to do with rules used to analyze data in archives written
+by CCX presentation team
+
+## Motivation
+
+Collecting data on some condition such as an alert is quite common pattern
+during the root-cause analysis and the ability to define the rules
+what to collect on which condition externally will help to do that faster.
+
+### Goals
+
+- Implement validation of the rules for conditional gatherer
+- Create a service taking gathering rules from the git repo and providing them through the API
+- Fetch the rules definitions from the service's API and apply them to the conditional gatherer
+
+### Non-Goals
+
+Empty
+
+## Proposal
+
+Right now we have conditional gatherer which can collect data based on some conditions
+which are defined in the code, for example the next config would collect 100 lines of logs
+from each container in namespace `openshift-cluster-samples-operator`
+and image streams of the same namespace when alert `SamplesImagestreamImportFailing` is firing
+and add it to the next archive.
+
+```json
+[{
+  "conditions": [{
+    "type": "alert_is_firing",
+    "params": {
+      "name": "SamplesImagestreamImportFailing"
+    }
+  }],
+  "gathering_functions": {
+    "logs_of_namespace": {
+      "namespace": "openshift-cluster-samples-operator",
+      "tail_lines": 100
+    },
+    "image_streams_of_namespace": {
+      "namespace": "openshift-cluster-samples-operator"
+    }
+  }
+}]
+```
+
+Conditions can have type and parameters, implemented types are:
+
+- `alert_is_firing` which is met when the alert with the name from parameter `name` is firing
+
+A gathering function consists of its name and parameters, implemented functions are:
+
+- `logs_of_namespace` which collects logs from each container in the namespace from parameter `namespace`
+limiting it to only last N lines from parameter `tail_lines`
+- `image_streams_of_namespace` which collects image streams of the namespace from parameter `namespace`
+
+The proposal is to implement the next process of adding new rules:
+
+1. We'll have a repo with json configs defining the rules. The repo is going to have a simple CI with validation
+against JSON schema and possibly some review process. The repo should live in https://github.com/RedHatInsights
+We have created a PoC for the repo:
+
+https://github.com/tremes/insights-operator-gathering-rules
+
+The JSON schema can be found here:
+
+https://raw.githubusercontent.com/openshift/insights-operator/f9b762149cd10ec98079e48b8a96fc02a2aca3c6/pkg/gatherers/conditional/gathering_rule.schema.json
+
+2. There will be a service living in console.redhat.com which will have the rules baked into the container and will
```

Comment:
is there precedence for this in the insights operator already (routinely calling back to console.redhat.com to get some config data?)

Reply:
Config - not, but we're getting the aggregator's data (https://cloud.redhat.com/api/insights-results-aggregator/v1/clusters/%s/report)

Reply:
In other words we're getting an Insights report (from console.redhat.com) for each uploaded IO archive to provide basic Insights overview for the Insights Advisor in the OCP console.